File: gateway/src/adapters/discord.ts
/** Discord adapter using discord.js. Handles DMs, mentions, and slash commands. */

import {
  Client,
  GatewayIntentBits,
  Message,
  Partials,
  type ChatInputCommandInteraction,
  type EmbedBuilder,
} from "discord.js";
import type { ChannelAdapter, MessageHandler, SessionChecker } from "./base.js";
import type { OutboundMessage } from "../models.js";
import { MessageType, type InboundMessage } from "../models.js";
import { registerSlashCommands } from "../slash-commands.js";

export interface DiscordConfig {
  botToken: string;
  botId?: string;
}

export class DiscordAdapter implements ChannelAdapter {
  readonly name = "discord";
  private client: Client;
  private config: DiscordConfig;
  private handler: MessageHandler | null = null;
  private resolvedBotId: string = "";
  private hasActiveSession: SessionChecker = () => false;

  /** Slash command interaction handler (set by Gateway). */
  onSlashCommand?: (
    command: string,
    options: Record<string, string>,
    senderId: string,
    senderName: string,
    chatId: string,
  ) => Promise<{ text?: string; embed?: EmbedBuilder }>;

  constructor(config: DiscordConfig) {
    this.config = config;
    this.client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.DirectMessages,
        GatewayIntentBits.MessageContent,
      ],
      partials: [Partials.Channel],
    });
  }

  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.client.once("ready", async () => {
        this.resolvedBotId = this.client.user?.id || this.config.botId || "";
        console.log(`[Discord] Connected as ${this.client.user?.tag} (${this.resolvedBotId})`);

        // Register slash commands
        try {
          await registerSlashCommands(this.config.botToken, this.resolvedBotId);
        } catch (err: any) {
          console.error(`[Discord] Failed to register slash commands: ${err.message}`);
        }

        resolve();
      });

      // Text message handler
      this.client.on("messageCreate", async (msg: Message) => {
        if (!this.handler) return;
        const parsed = this.parseMessage(msg);
        if (parsed) await this.handler(parsed);
      });

      // Slash command handler
      this.client.on("interactionCreate", async (interaction) => {
        if (!interaction.isChatInputCommand()) return;
        await this.handleSlashCommand(interaction);
      });

      this.client.login(this.config.botToken).catch(reject);
    });
  }

  async disconnect(): Promise<void> {
    await this.client.destroy();
  }

  async sendMessage(message: OutboundMessage): Promise<string | undefined> {
    const channel = await this.client.channels.fetch(message.chatId);
    if (!channel || !channel.isTextBased() || !("send" in channel)) return undefined;

    const sendable = channel as any;

    // Edit existing message if editMessageId is set
    if (message.editMessageId) {
      try {
        const existing = await sendable.messages.fetch(message.editMessageId);
        if (existing) {
          const payload = message.embed
            ? { embeds: [message.embed], content: "" }
            : { content: message.text, embeds: [] };
          await existing.edit(payload);
          return message.editMessageId;
        }
      } catch {
        // Message might have been deleted, fall through to send new
      }
    }

    // Send embed if provided
    if (message.embed) {
      const sent = await sendable.send({ embeds: [message.embed] });
      return sent?.id;
    }

    // Plain text with smart splitting at newlines
    const text = message.text;
    if (text.length <= 2000) {
      const sent = await sendable.send(text);
      return sent?.id;
    } else {
      let lastId: string | undefined;
      for (const chunk of smartSplit(text, 2000)) {
        const sent = await sendable.send(chunk);
        lastId = sent?.id;
      }
      return lastId;
    }
  }

  registerHandler(handler: MessageHandler): void {
    this.handler = handler;
  }

  setSessionChecker(checker: SessionChecker): void {
    this.hasActiveSession = checker;
  }

  private async handleSlashCommand(interaction: ChatInputCommandInteraction): Promise<void> {
    if (!this.onSlashCommand) {
      await interaction.reply({ content: "Commands not ready yet.", ephemeral: true });
      return;
    }

    const options: Record<string, string> = {};
    for (const opt of interaction.options.data) {
      options[opt.name] = String(opt.value ?? "");
    }

    try {
      await interaction.deferReply();

      const result = await this.onSlashCommand(
        interaction.commandName,
        options,
        interaction.user.id,
        interaction.user.displayName || interaction.user.username,
        interaction.channelId,
      );

      if (result.embed) {
        await interaction.editReply({ embeds: [result.embed] });
      } else {
        await interaction.editReply({ content: result.text || "Done." });
      }
    } catch (err: any) {
      const content = `Error: ${err.message}`;
      try {
        await interaction.editReply({ content });
      } catch {
        // Interaction may have expired
      }
    }
  }

  private parseMessage(msg: Message): InboundMessage | null {
    if (msg.author.bot) return null;

    const isDM = !msg.guild;
    const isMentioned = msg.mentions.has(this.resolvedBotId);
    const hasSession = this.hasActiveSession(msg.author.id);


    // Also treat literal @BotName as a mention (typed manually, not via picker)
    const botName = this.client.user?.displayName || this.client.user?.username || "";
    const isLiteralMention = botName
      ? new RegExp(`^@${botName}\\b`, "i").test(msg.content)
      : false;

    if (!isDM && !isMentioned && !isLiteralMention && !hasSession) return null;

    let text = msg.content;
    // Strip Discord mention format: <@botId>, <@!botId>, and <@&roleId>
    if (isMentioned) {
      text = text.replace(/<@[!&]?[\w-]+>/g, "").trim();
    }
    // Strip literal @BotName prefix (typed manually, not via mention picker)
    if (botName) {
      text = text.replace(new RegExp(`^@${botName}\\b`, "i"), "").trim();
    }

    if (!text) return null;

    return {
      channel: "discord",
      chatId: msg.channel.id,
      senderId: msg.author.id,
      senderName: msg.author.displayName || msg.author.username,
      text,
      messageType: MessageType.TEXT,
      timestamp: msg.createdAt,
      raw: {},
    };
  }
}

/** Split text at newline boundaries instead of mid-word. Exported for testing. */
export function smartSplit(text: string, maxLen: number): string[] {
  const chunks: string[] = [];
  let remaining = text;

  while (remaining.length > maxLen) {
    let splitAt = remaining.lastIndexOf("\n", maxLen);
    if (splitAt <= 0) splitAt = maxLen;
    chunks.push(remaining.slice(0, splitAt));
    remaining = remaining.slice(splitAt).replace(/^\n/, "");
  }

  if (remaining) chunks.push(remaining);
  return chunks;
}

File: gateway/adapters/discord.py
"""Discord adapter using webhook receive + REST API send."""

from __future__ import annotations

import logging
from typing import Optional

import httpx

from gateway.base import ChannelAdapter, MessageHandler
from gateway.models import InboundMessage, OutboundMessage, MessageType

logger = logging.getLogger(__name__)

_DISCORD_API_BASE = "https://discord.com/api/v10"
_TIMEOUT = 15


class DiscordAdapter(ChannelAdapter):
    """Discord integration via webhook receive and REST API send."""

    def __init__(
        self,
        bot_token: str,
        bot_id: Optional[str] = None,
        guild_id: Optional[str] = None,
    ):
        self._bot_token = bot_token
        self._bot_id = bot_id
        self._guild_id = guild_id
        self._handler: Optional[MessageHandler] = None

    @property
    def name(self) -> str:
        return "discord"

    async def connect(self) -> None:
        """No-op in webhook mode -- Discord pushes events to our endpoint."""
        pass

    async def disconnect(self) -> None:
        """No-op in webhook mode."""
        pass

    async def send_message(self, message: OutboundMessage) -> None:
        """Send a text message via Discord REST API."""
        url = f"{_DISCORD_API_BASE}/channels/{message.chat_id}/messages"
        headers = {
            "Authorization": f"Bot {self._bot_token}",
            "Content-Type": "application/json",
        }
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(url, json={"content": message.text}, headers=headers)
            if resp.status_code >= 400:
                logger.error(
                    "Failed to send Discord message: %s %s",
                    resp.status_code,
                    resp.text[:200],
                )
            resp.raise_for_status()

    async def register_handler(self, handler: MessageHandler) -> None:
        """Register the message handler. Called by the gateway on startup."""
        self._handler = handler

    def parse_webhook(self, payload: dict) -> InboundMessage | None:
        """Parse a Discord Gateway dispatch payload (MESSAGE_CREATE).

        Expected format:
        {
            "t": "MESSAGE_CREATE",
            "d": {
                "id": "...",
                "channel_id": "...",
                "guild_id": "...",   # absent for DMs
                "author": {"id": "...", "username": "...", "bot": false},
                "content": "hello",
                "type": 0
            }
        }
        """
        if payload.get("t") != "MESSAGE_CREATE":
            return None

        d = payload.get("d", {})
        if not d:
            return None

        if d.get("author", {}).get("bot", False):
            return None

        if d.get("type", 0) != 0:
            return None

        content = d.get("content", "").strip()
        if not content:
            return None

        if self._bot_id:
            prefix = f"<@{self._bot_id}>"
            if content.startswith(prefix):
                content = content[len(prefix):].strip()

        return InboundMessage(
            channel="discord",
            chat_id=d["channel_id"],
            sender_id=d["author"]["id"],
            sender_name=d["author"]["username"],
            text=content,
            message_type=MessageType.TEXT,
            raw=payload,
        )

    async def handle_webhook(self, payload: dict) -> None:
        """Called by the webhook endpoint. Parses and dispatches to handler."""
        message = self.parse_webhook(payload)
        if message and self._handler:
            await self._handler(message)
